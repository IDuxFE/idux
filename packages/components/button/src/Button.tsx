/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VNodeTypes } from 'vue'

import { computed, defineComponent, inject, normalizeClass } from 'vue'

import { useGlobalConfig } from '@idux/components/config'
import { FORM_TOKEN } from '@idux/components/form'
import { IxIcon } from '@idux/components/icon'

import { buttonToken } from './token'
import { buttonProps } from './types'

export default defineComponent({
  name: 'IxButton',
  props: buttonProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-button`)
    const config = useGlobalConfig('button')
    const groupProps = inject(buttonToken, {})
    const formContext = inject(FORM_TOKEN, null)

    const mode = computed(() => props.mode ?? groupProps.mode ?? 'default')
    const size = computed(() => props.size ?? groupProps.size ?? formContext?.size.value ?? config.size)

    const classes = computed(() => {
      const { block, danger, disabled, ghost, loading, icon, shape = groupProps.shape } = props
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [prefixCls]: true,
        [`${prefixCls}-block`]: block,
        [`${prefixCls}-danger`]: danger,
        [`${prefixCls}-disabled`]: disabled || loading,
        [`${prefixCls}-ghost`]: ghost,
        [`${prefixCls}-loading`]: loading,
        [`${prefixCls}-icon-only`]: !slots.default && (icon || loading),
        [`${prefixCls}-${mode.value}`]: mode.value !== 'default',
        [`${prefixCls}-${shape}`]: !!shape,
        [`${prefixCls}-${size.value}`]: true,
      })
    })

    return () => {
      const { disabled, loading, icon, type } = props

      const children: VNodeTypes[] = []
      if (loading) {
        children.push(<IxIcon name="loading"></IxIcon>)
      } else if (slots.icon) {
        children.push(slots.icon())
      } else if (icon) {
        children.push(<IxIcon name={icon}></IxIcon>)
      }
      if (slots.default) {
        children.push(<span>{slots.default()}</span>)
      }

      if (mode.value === 'link') {
        return <a class={classes.value}>{children}</a>
      }
      return (
        <button class={classes.value} disabled={disabled || loading} type={type}>
          {children}
        </button>
      )
    }
  },
})
