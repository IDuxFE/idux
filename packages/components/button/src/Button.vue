<template>
  <component :is="tag" class="ix-button" :class="classes" :disabled="disabled || loading ? true : undefined">
    <ix-icon v-if="loading" name="loading" />
    <ix-icon v-else-if="icon" :name="icon" />
    <span v-if="hasDefaultSlot"><slot></slot></span>
  </component>
</template>

<script lang="ts">
import type { ComputedRef, Ref } from 'vue'
import type { ButtonConfig } from '@idux/components/core/config'
import type { ButtonMode } from '@idux/components/core/types'
import type { ButtonGroupProps, ButtonProps } from './types'

import { computed, defineComponent, inject, onUpdated, ref } from 'vue'
import { PropTypes } from '@idux/cdk/utils'
import { useGlobalConfig } from '@idux/components/core/config'
import { IxIcon } from '@idux/components/icon'
import { buttonGroupInjectionKey } from './button'

export default defineComponent({
  name: 'IxButton',
  components: { IxIcon },
  props: {
    mode: PropTypes.oneOf(['primary', 'default', 'dashed', 'text', 'link'] as const),
    danger: PropTypes.bool,
    ghost: PropTypes.bool,
    disabled: PropTypes.bool,
    loading: PropTypes.bool,
    size: PropTypes.oneOf(['large', 'medium', 'small'] as const),
    shape: PropTypes.oneOf(['circle', 'round'] as const),
    block: PropTypes.bool,
    icon: PropTypes.string,
  },
  setup(props: ButtonProps, { slots }) {
    const groupProps = inject(buttonGroupInjectionKey, {})
    const buttonConfig = useGlobalConfig('button')

    const mode = computed(() => props.mode ?? (groupProps.mode || buttonConfig.mode))

    const hasDefaultSlot = ref(!!slots.default)
    onUpdated(() => {
      hasDefaultSlot.value = !!slots.default
    })
    const classes = useClasses(props, groupProps, buttonConfig, mode, hasDefaultSlot)

    const tag = computed(() => (mode.value === 'link' ? 'a' : 'button'))

    return { classes, tag, hasDefaultSlot }
  },
})

const useClasses = (
  props: ButtonProps,
  groupProps: ButtonGroupProps,
  config: ButtonConfig,
  mode: ComputedRef<ButtonMode>,
  hasDefaultSlot: Ref<boolean>,
) => {
  return computed(() => {
    const size = props.size ?? (groupProps.size || config.size)
    const shape = props.shape ?? groupProps.shape
    return [
      mode.value !== 'default' ? `ix-button-${mode.value}` : '',
      size !== 'medium' ? `ix-button-${size}` : '',
      shape ? `ix-button-${shape}` : '',
      {
        'ix-button-danger': props.danger,
        'ix-button-ghost': props.ghost,
        'ix-button-disabled': props.disabled,
        'ix-button-loading': props.loading,
        'ix-button-block': props.block,
        'ix-button-icon-only': !hasDefaultSlot.value && (!!props.icon || props.loading),
      },
    ]
  })
}
</script>
