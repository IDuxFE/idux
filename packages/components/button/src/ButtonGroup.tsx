/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import { computed, defineComponent, provide } from 'vue'

import { useGlobalConfig } from '@idux/components/config'

import { buttonToken } from './token'
import { buttonGroupProps } from './types'

export default defineComponent({
  name: 'IxButtonGroup',
  props: buttonGroupProps,
  setup(props, { slots }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-button-group`)

    provide(buttonToken, props)

    return () => <div class={mergedPrefixCls.value}>{slots.default && slots.default()}</div>
  },
})
