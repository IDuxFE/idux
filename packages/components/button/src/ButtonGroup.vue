<template>
  <div class="ix-button-group">
    <slot></slot>
  </div>
</template>

<script lang="ts">
import { defineComponent, provide } from 'vue'
import { PropTypes } from '@idux/cdk/utils'
import { buttonGroupInjectionKey } from './button'

import type { ButtonGroupProps } from './types'

export default defineComponent({
  name: 'IxButtonGroup',
  props: {
    mode: PropTypes.oneOf(['primary', 'default', 'dashed', 'text', 'link'] as const),
    size: PropTypes.oneOf(['large', 'medium', 'small'] as const),
    shape: PropTypes.oneOf(['circle', 'round'] as const),
  },
  setup(props: ButtonGroupProps) {
    provide(buttonGroupInjectionKey, props)
  },
})
</script>
