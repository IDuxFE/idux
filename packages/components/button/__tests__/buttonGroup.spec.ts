import { mount } from '@vue/test-utils'

import { renderWork } from '@tests'

import Button from '../src/Button'
import ButtonGroup from '../src/ButtonGroup'

const TestComponent = {
  components: { ButtonGroup, Button },
  template: `
  <ButtonGroup :mode="mode" :size="size" :shape="shape">
    <Button>default</Button>
    <Button mode="primary">primary</Button>
    <Button size="lg">large</Button>
    <Button shape="circle">circle</Button>
  </ButtonGroup>
  `,
  props: {
    mode: String,
    size: String,
    shape: String,
  },
}

describe('ButtonGroup', () => {
  renderWork(TestComponent)

  test('mode work', async () => {
    const wrapper = mount(TestComponent)

    expect(wrapper.findAll('.ix-button-primary').length).toBe(1)
    expect(wrapper.findAll('.ix-button-dashed').length).toBe(0)

    await wrapper.setProps({ mode: 'dashed' })

    expect(wrapper.findAll('.ix-button-primary').length).toBe(1)
    expect(wrapper.findAll('.ix-button-dashed').length).toBe(3)
  })

  test('size work', async () => {
    const wrapper = mount(TestComponent)

    expect(wrapper.findAll('.ix-button-lg').length).toBe(1)
    expect(wrapper.findAll('.ix-button-md').length).toBe(3)

    await wrapper.setProps({ size: 'xl' })

    expect(wrapper.findAll('.ix-button-lg').length).toBe(1)
    expect(wrapper.findAll('.ix-button-xl').length).toBe(3)
  })

  test('shape work', async () => {
    const wrapper = mount(TestComponent)

    expect(wrapper.findAll('.ix-button-circle').length).toBe(1)
    expect(wrapper.findAll('.ix-button-round').length).toBe(0)

    await wrapper.setProps({ shape: 'round' })

    expect(wrapper.findAll('.ix-button-circle').length).toBe(1)
    expect(wrapper.findAll('.ix-button-round').length).toBe(3)
  })
})
