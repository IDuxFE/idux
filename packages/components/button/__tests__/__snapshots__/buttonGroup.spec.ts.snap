// Vitest Snapshot v1

exports[`ButtonGroup > render work 1`] = `"<div class=\\"ix-button-group\\"><button class=\\"ix-button ix-button-md\\" type=\\"button\\"><span>default</span></button><button class=\\"ix-button ix-button-primary ix-button-md\\" type=\\"button\\"><span>primary</span></button><button class=\\"ix-button ix-button-lg\\" type=\\"button\\"><span>large</span></button><button class=\\"ix-button ix-button-circle ix-button-md\\" type=\\"button\\"><span>circle</span></button></div>"`;
